File: HW1/task1.py
import sys
import json
from pyspark import SparkContext, SparkConf
import pyspark
from operator import add


sc= SparkContext("local[*]")
sc.setLogLevel("ERROR")

file_path= sys.argv[1]
output_path= sys.argv[2]
stopwords = sys.argv[3]
y = sys.argv[4]
m = int(sys.argv[5])
n = int(sys.argv[6])


punctuations = ['(',')','[',']',',','.','!','?',':',';']
res = {}


with open(stopwords, 'r') as f:
    myNames = [line.strip() for line in f]
    
#print(myNames)

#print("-------------")
#TASK 1
#print("TASK1")
reviews = sc.textFile(file_path)
total_reviews = reviews.map(lambda x: json.loads(x))
total_reviews_count = total_reviews.count()
res['A'] = total_reviews_count


#TASK 2
#print("TASK2")
task2 = total_reviews.filter(lambda x: x['date'].split(' ')[0].split('-')[0] == y)
reviews_year = task2.count()
res['B'] = reviews_year
#print(reviews_year)

#TASK 3
#print("TASK3")
users = total_reviews.map(lambda x: x['user_id']).distinct()
distinct_users = users.count()
res['C'] = distinct_users
#print(distinct_users)

#TASK 4
#print("TASK4")
top_users = total_reviews.map(lambda x : (x['user_id'],1)).reduceByKey(lambda x,y : x + y).sortBy(lambda x: (-x[1],x[0])).take(m)
res['D'] = top_users

#with open(output_path,'w') as f:
#    json.dump(res , f)

        
def removepunctuation(word):
    word = word.strip()
    for char in word:
        if char in punctuations:
            word = word.replace(char, "")
    return (word.lower())


        
#TASK 5
#print("TASK5")
words = total_reviews.flatMap(lambda x : x['text'].strip().split(' '))
words = words.map(removepunctuation).filter(lambda x: x not in myNames).filter(lambda x: x!= "")
words  = words.map(lambda x : (x,1))
#print("1")
awords =  words.reduceByKey(lambda x,y : x + y)
#print("2")
sorted_word_counts = awords.sortBy(lambda x: (-x[1], x[0]))
#print("3")
top_n_freq_words = sorted_word_counts.map(lambda x: x[0])
res['E'] = top_n_freq_words.take(n)
#print(res['E'])
#print(top_n_freq_words.take(20))
#print("Done")


with open(output_path,'w') as f:
    json.dump(res , f)








    

File: HW1/src/main/scala/task1.scala
import java.io.{File, PrintWriter}
import org.json4s._
import org.json4s.jackson.JsonMethods._
import org.apache.log4j.{Level, Logger}
import org.apache.spark.sql.SparkSession
import scala.collection.immutable.ListMap
import scala.io.Source
import java.io.FileWriter
import scala.collection.mutable.ListBuffer




object task1 {
  val punctuations = List('(',')','[',']',',','.','!','?',':',';')
  def main(args: Array[String]): Unit = {
    Logger.getLogger("org").setLevel(Level.ERROR)


    val spark = SparkSession.builder
      .master("local[*]")
      .appName("inf-553_task1")
      .getOrCreate()

    val input_file = args(0)
    val output_file = args(1)
    val stopwords = args(2)
    val y = args(3)
    val m = args(4).toInt
    val n = args(5).toInt



    val sc = spark.sparkContext;

    val data = spark.read.json(input_file).rdd

    // Useful reviews
    val total_reviews = data.map(x => (x.getAs[String]("review_id"))).count()
//    println(total_reviews)

    val year_reviews = data.map(x => (x.getAs[String]("date"))).filter(x => x.contains(y)).count()
//    println(year_reviews)

    val distinct_users =  data.map(x => (x.getAs[String]("user_id"))).distinct().count()
//    println(distinct_users)

    val top_m_users = data.map(x => (x.getAs[String]("user_id"),1)).reduceByKey((x,y) => x+y).sortBy(x => (-x._2, x._1)).take(m).toList
//    println(top_m_users.toList)

    var words_to_remove = ListBuffer[String]()
    for (line <- Source.fromFile(stopwords).getLines) {
      val lines  = line.trim
      words_to_remove+=lines
    }
//    println(words_to_remove)
    val slist = words_to_remove.toList
//    println(slist)



    val words = data.flatMap(x => (x.getAs[String]("text")).toLowerCase.trim.split(" ")).map(x => x.replaceAll("\\p{Punct}", "")).filter(!slist.contains(_)).filter(x=> x!="").map(x => (x,1)).reduceByKey((x,y) => x+y).sortBy(x => (-x._2, x._1)).keys.take(n).toList
//    println(words.keys.take(10).toList)
//    words.map(x => (x.getAs[String]("user_id"),1)).reduceByKey((x,y) => x+y).sortBy(x => (-x._2, x._1)).take(10)
//    println(words.toList)


    val file = new FileWriter(output_file)
    file.write("{")

    file.write("\"A\": " + total_reviews + ",")

    file.write("\"B\": " + year_reviews + ",")

    file.write("\"C\": " + distinct_users + ",")

    file.write("\"D\": [")
    var i = 1
    for ((k, v) <- top_m_users) {
      file.write("[")
      file.write("\""+k+"\"")
      file.write("," + v)
      i = i+1
      if (i == m+1) {
        file.write("]")
      } else {
        file.write("],")
      }
    }
    file.write("],")

    file.write("\"E\": [")
    var j=1
    for (x <- words) {

      file.write("\""+x+"\"")
      j=j+1
      if(j==n+1)
        {
          file.write("]")
        }
      else{
        file.write(",")
      }

    }
    file.write("}")
    file.flush()


























  }

  def removep(word:String) :String={
    val new_word2 = new StringBuilder("")
    val word = " Arpit [Goel "
    val new_word = word.trim()
//    val punctuations = List('(',')','[',']',',','.','!','?',':',';')
    print(new_word)
    for(c <- word)
    {
      if(!punctuations.contains(c))
      {
        new_word2.append(c)
      }
    }
    return(new_word2.toString())

  }


}
